pkg/cloud/services/loadbalancer/loadbalancer.go

```diff
@@ -175,6 +186,70 @@ func (s *Service) getOrCreateListener(openStackCluster *infrav1.OpenStackCluster
 	return listener, nil
 }
 
+func (s *Service) getOrUpdateAllowedCIDRS(openStackCluster *infrav1.OpenStackCluster, listener *listeners.Listener) error {
+	allowedCIDRs := []string{}
+
+	if len(openStackCluster.Spec.APIServerLoadBalancer.AllowedCIDRs) > 0 {
+		allowedCIDRs = append(allowedCIDRs, openStackCluster.Spec.APIServerLoadBalancer.AllowedCIDRs...)
+
+		if openStackCluster.Spec.Bastion.Enabled {
+			allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Bastion.FloatingIP, openStackCluster.Status.Bastion.IP)
+		}
+
+		if openStackCluster.Status.Network.Subnet.CIDR != "" {
+			allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Network.Subnet.CIDR)
+		}
+
+		if len(openStackCluster.Status.Network.Router.IPs) > 0 {
+			allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Network.Router.IPs...)
+		}
+	}
+
+	// validate CIDRs and convert any given IP into a CIDR
+	allowedCIDRs = validateIPs(allowedCIDRs)
+
+	// sort and remove duplicates
+	allowedCIDRs = capostrings.Unique(allowedCIDRs)
+	listener.AllowedCIDRs = capostrings.Unique(listener.AllowedCIDRs)
+
+	if !reflect.DeepEqual(allowedCIDRs, listener.AllowedCIDRs) {
+		listenerUpdateOpts := listeners.UpdateOpts{
+			AllowedCIDRs: &allowedCIDRs,
+		}
+
+		listener, err := s.loadbalancerClient.UpdateListener(listener.ID, listenerUpdateOpts)
+		if err != nil {
+			record.Warnf(openStackCluster, "FailedUpdateListener", "Failed to update listener %s: %v", listener.Name, err)
+			return err
+		}
+
+		if err := s.waitForListener(listener.ID, "ACTIVE"); err != nil {
+			record.Warnf(openStackCluster, "FailedUpdateListener", "Failed to update listener %s with id %s: wait for listener active: %v", listener.Name, listener.ID, err)
+			return err
+		}
+
+		record.Eventf(openStackCluster, "SuccessfulUpdateListener", "Updated allowed_cidrs %s for listener %s with id %s", listener.AllowedCIDRs, listener.Name, listener.ID)
+
+		return nil
+	}
+	return nil
+}
+
+// validateIPs validate given IPs/CIDRs and remove non valid network objects.
+func validateIPs(definedCIDRs []string) []string {
+	marshaledCIDRs := []string{}
+
+	for _, v := range definedCIDRs {
+		if net.IsIPv4String(v) {
+			marshaledCIDRs = append(marshaledCIDRs, v+"/32")
+		} else if net.IsIPv4CIDRString(v) {
+			marshaledCIDRs = append(marshaledCIDRs, v)
+		}
```

Comment:
If one entry doesn't match the type ipv4 nor cidr then it will just be ignored without any warning. Should we catch cases where the string doesn't match?

Reply:
if we stay on the current implementation (no input validation during Create/Update/Delete on `spec.APIServerLoadBalancer.AllowedCIDRs`) we should at least log if a `string` got dropped due invalid ipv4 or cidr.

But yes, i'm open for alternative ideas (but as we don't have any existing input validation i decided to go that way).